packages/agreement-process/src/model/domain/validators.ts

```diff
@@ -0,0 +1,341 @@
+    throw agreementNotInExpectedState(agreementId, agreementState);
+  }
+};
+export const certifiedAttributesSatisfied = (
+  descriptor: Descriptor,
+  tenant: Tenant
+): boolean => {
+  const certifiedAttributes = tenant.attributes.filter(
+    (e) => e.type === tenantAttributeType.CERTIFIED
+  ) as CertifiedTenantAttribute[];
+
+  return attributesSatisfied(
+    descriptor.attributes.certified,
+    certifiedAttributes
+  );
+};
+
+export const declaredAttributesSatisfied = (
+  descriptor: Descriptor,
+  tenant: Tenant
+): boolean => {
+  const declaredAttributes = tenant.attributes.filter(
+    (e) => e.type === tenantAttributeType.DECLARED
+  ) as DeclaredTenantAttribute[];
+
+  return attributesSatisfied(
+    descriptor.attributes.declared,
+    declaredAttributes
+  );
+};
+
+export const verifiedAttributesSatisfied = (
+  agreement: Agreement,
+  descriptor: Descriptor,
+  tenant: Tenant
+): boolean => {
+  const verifiedAttributes = tenant.attributes.filter(
+    (e) => e.type === tenantAttributeType.DECLARED
+  ) as VerifiedTenantAttribute[];
+
+  const producersAttributesNotExpired = verifiedAttributes.filter(
+    (a): boolean =>
+      !!a.verifiedBy.find(
+        (v) =>
+          v.id === agreement.producerId &&
+          (!v.extensionDate || v.extensionDate > new Date())
+      )
+  );
+
+  return attributesSatisfied(
+    descriptor.attributes.verified,
+    producersAttributesNotExpired
+  );
+};
+
+const matchingAttributes = (
+  eServiceAttributes: EServiceAttribute[][],
+  consumerAttributes: string[]
+): string[] =>
+  eServiceAttributes
+    .flatMap((atts) => atts.map((att) => att.id))
+    .filter((att) => consumerAttributes.includes(att));
+
+export const matchingCertifiedAttributes = (
+  descriptor: Descriptor,
+  consumer: Tenant
+): CertifiedAgreementAttribute[] => {
+  const attributes = consumer.attributes
+    .filter(
+      (a) => a.type === tenantAttributeType.CERTIFIED && !a.revocationTimestamp
+    )
+    .map((a) => a.id);
+
+  return matchingAttributes(descriptor.attributes.certified, attributes).map(
+    (id) => ({
+      type: agreementAttributeType.CERTIFIED,
+      id,
+    })
+  );
+};
+
+export const matchingDeclaredAttributes = (
+  descriptor: Descriptor,
+  consumer: Tenant
+): DeclaredAgreementAttribute[] => {
+  const attributes = consumer.attributes
+    .filter(
+      (a) => a.type === tenantAttributeType.DECLARED && !a.revocationTimestamp
+    )
+    .map((a) => a.id);
+
+  return matchingAttributes(descriptor.attributes.declared, attributes).map(
+    (id) => ({
+      type: agreementAttributeType.DECLARED,
+      id,
+    })
+  );
+};
+
+export const matchingVerifiedAttributes = (
+  eService: EService,
+  descriptor: Descriptor,
+  consumer: Tenant
+): VerifiedAgreementAttribute[] => {
+  const attributes = consumer.attributes
+    .filter(
+      (a) =>
+        a.type === tenantAttributeType.VERIFIED &&
+        a.verifiedBy.find((v) => v.id === eService.producerId)
```

Comment:
we need to verify the validity date as done [here](https://github.com/pagopa/interop-be-monorepo/pull/115/files#diff-a884fb59619b9a28ddc25a8f7e1b048b7ebb02f1e09b472f6ff3f4aa9cf7db9bR232).
Maybe we can create a function to reuse?

Note: this check is not performed in the as-is and it's a bug, cc @beetlecrunch .